api/src/indigo_structure_checker.cpp

```diff
@@ -0,0 +1,205 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ***************************************************************************/
+
+#include "indigo_structure_checker.h"
+#include "api/indigo.h"
+#include "api/src/indigo_molecule.h"
+#include "api/src/indigo_reaction.h"
+#include "base_cpp/tlscont.h"
+#include "molecule/molecule_automorphism_search.h"
+#include "third_party/rapidjson/stringbuffer.h"
+#include "third_party/rapidjson/writer.h"
+#include <functional>
+#include <regex>
+#include <string>
+
+using namespace indigo;
+
+struct CheckParams
+{
+    int check_flags = StructureChecker2::CHECK_ALL;
+    std::vector<int> selected_atoms;
+    std::vector<int> selected_bonds;
+};
+
+static constexpr const char* checkTypeName[] = {
+    IndigoStructureChecker::CHECK_NONE_TXT,      IndigoStructureChecker::CHECK_LOAD_TXT,         IndigoStructureChecker::CHECK_VALENCE_TXT,
+    IndigoStructureChecker::CHECK_RADICAL_TXT,   IndigoStructureChecker::CHECK_PSEUDOATOM_TXT,   IndigoStructureChecker::CHECK_STEREO_TXT,
+    IndigoStructureChecker::CHECK_QUERY_TXT,     IndigoStructureChecker::CHECK_OVERLAP_ATOM_TXT, IndigoStructureChecker::CHECK_OVERLAP_BOND_TXT,
+    IndigoStructureChecker::CHECK_RGROUP_TXT,    IndigoStructureChecker::CHECK_SGROUP_TXT,       IndigoStructureChecker::CHECK_TGROUP_TXT,
+    IndigoStructureChecker::CHECK_CHIRALITY_TXT, IndigoStructureChecker::CHECK_CHIRAL_FLAG_TXT,  IndigoStructureChecker::CHECK_3D_COORD_TXT,
+    IndigoStructureChecker::CHECK_CHARGE_TXT,    IndigoStructureChecker::CHECK_SALT_TXT,         IndigoStructureChecker::CHECK_AMBIGUOUS_H_TXT,
+    IndigoStructureChecker::CHECK_COORD_TXT,     IndigoStructureChecker::CHECK_V3000_TXT,        IndigoStructureChecker::CHECK_ALL_TXT};
+
+static CheckParams check_params_from_string(const char* params)
+{
+    CheckParams r;
+    size_t len = sizeof(checkTypeName) / sizeof(*checkTypeName);
+    if (params)
+    {
+        std::smatch sm1;
+        std::unordered_set<std::string> words;
+        std::string s(params);
+        std::regex rx1(R"(\b(\w+)\b)", std::regex_constants::icase);
+        while (std::regex_search(s, sm1, rx1))
+        {
+            words.insert(sm1[1]);
+            s = sm1.suffix();
+        }
+        r.check_flags = StructureChecker2::CHECK_NONE;
+        for (int i = 0; i < len; i++)
+        {
+            if (words.find(std::string(checkTypeName[i])) != words.end())
+            {
+                r.check_flags |= 1 << i;
```

Comment:
Overcomplicate logic. In order to add a new type to check one need to go to 4 places and append correct bitwise value. Instead, it is suggested to have mapping between codes and strings (see comment below)